thrift/error.go

```diff
@@ -0,0 +1,111 @@
+package thrift
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+)
+
+type MissingField struct {
+	Field Field
+}
+
+func (e *MissingField) Error() string {
+	return fmt.Sprintf("missing required field: %s", e.Field)
+}
+
+type TypeMismatch struct {
+	Expect Type
+	Found  Type
+	item   string
+}
+
+func (e *TypeMismatch) Error() string {
+	return fmt.Sprintf("%s type mismatch: expected %s but found %s", e.item, e.Expect, e.Found)
+}
+
+type decodeError struct {
+	base error
+	path []error
+}
+
+func (e *decodeError) Error() string {
+	s := strings.Builder{}
+	s.Grow(256)
+	s.WriteString("decoding thrift payload: ")
+
+	if len(e.path) != 0 {
+		n := len(e.path) - 1
+		for i := n; i >= 0; i-- {
+			if i < n {
+				s.WriteString(" → ")
+			}
+			s.WriteString(e.path[i].Error())
+		}
+		s.WriteString(": ")
+	}
+
+	s.WriteString(e.base.Error())
+	return s.String()
+}
+
+func (e *decodeError) Unwrap() error { return e.base }
+
+func with(base, elem error) error {
+	if errors.Is(base, io.EOF) {
+		return base
+	}
+	e, _ := base.(*decodeError)
+	if e == nil {
+		e = &decodeError{base: base}
+	}
+	e.path = append(e.path, elem)
+	return e
+}
+
+type decodeErrorField struct {
+	field Field
+}
+
+func (d *decodeErrorField) Error() string {
+	return d.field.String()
+}
+
+type decodeErrorList struct {
+	list  List
+	index int
+}
+
+func (d *decodeErrorList) Error() string {
+	return fmt.Sprintf("%d/%d:%s", d.index, d.list.Size, d.list)
+}
+
+type decodeErrorSet struct {
+	set   Set
+	index int
+}
+
+func (d *decodeErrorSet) Error() string {
+	return fmt.Sprintf("%d/%d:%s", d.index, d.set.Size, d.set)
+}
+
+type decodeErrorMap struct {
+	_map  Map
```

Comment:
Why the `_` prefix?

Reply:
Isn't `map` a keyword in Go?

Reply:
Sorry I meant as opposed to a name like `errors`.

Reply:
This is the map value that caused the error, maybe we call it `cause`? Would need to change the other error types too for consistency.

Did you run this on a linter that complained?

Reply:
Nope, I thought it was the map of errors, so figured it was odd to resort to an underscore prefix while there was some easier naming. But I'm just seeing now it was the map that caused the error 😅 

+1 for `cause`.